src/main/java/org/fiware/iam/tmforum/handlers/ProductOrderEventHandler.java

```diff
@@ -0,0 +1,135 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.fiware.iam.til.TrustedIssuersListAdapter;
+import org.fiware.iam.tmforum.OrganizationResolver;
+import org.fiware.iam.tmforum.productorder.model.*;
+import org.fiware.rainbow.api.AgreementApiClient;
+import org.fiware.rainbow.model.AgreementCreateVO;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@RequiredArgsConstructor
+@Singleton
+@Slf4j
+public class ProductOrderEventHandler implements EventHandler {
+
+	private static final String CREATE_EVENT = "ProductOrderCreateEvent";
+	private static final String DELETE_EVENT = "ProductOrderDeleteEvent";
+	private static final String STATE_CHANGE_EVENT = "ProductOrderStateChangeEvent";
+
+	private static final List<String> SUPPORTED_EVENT_TYPES = List.of(CREATE_EVENT, DELETE_EVENT, STATE_CHANGE_EVENT);
+
+	private final ObjectMapper objectMapper;
+	private final OrganizationResolver organizationResolver;
+	private final TrustedIssuersListAdapter trustedIssuersListAdapter;
+	private final AgreementApiClient agreementApiClient;
+
+	@Override
+	public boolean isEventTypeSupported(String eventType) {
+		return SUPPORTED_EVENT_TYPES.contains(eventType);
+	}
+
+	@Override
+	public Mono<HttpResponse<?>> handleEvent(String eventType, Map<String, Object> event) {
+
+		String orgId = Stream
+				.ofNullable(event)
+				.map(rawEvent -> objectMapper.convertValue(rawEvent, ProductOrderCreateEventVO.class))
+				.map(ProductOrderCreateEventVO::getEvent)
+				.map(ProductOrderCreateEventPayloadVO::getProductOrder)
+				.map(ProductOrderVO::getRelatedParty)
+				.map(rpl -> {
+					if (rpl.size() > 1) {
+						throw new IllegalArgumentException("Expected exactly one ordering organization.");
+					}
+					return rpl.get(0);
+				})
+				.map(RelatedPartyVO::getId)
+				.findAny()
+				.orElseThrow(() -> new IllegalArgumentException("The ProductOrder-Event does not include a valid organization id."));
+
+		return switch (eventType) {
+			case CREATE_EVENT -> handelCreateEvent(orgId, event);
+			case STATE_CHANGE_EVENT -> handelStateChangeEvent(orgId, event);
+			case DELETE_EVENT -> handelDeleteEvent(orgId);
+			default -> throw new IllegalArgumentException("Invalid event type received.");
+		};
+
+	}
+
+	private Mono<HttpResponse<?>> handelCreateEvent(String organizationId, Map<String, Object> event) {
+		ProductOrderCreateEventVO productOrderCreateEventVO = objectMapper.convertValue(event, ProductOrderCreateEventVO.class);
+
+		ProductOrderVO productOrderVO = Optional.ofNullable(productOrderCreateEventVO.getEvent())
+				.map(ProductOrderCreateEventPayloadVO::getProductOrder)
+				.orElseThrow(() -> new IllegalArgumentException("The event does not contain a product order."));
+
+		boolean isCompleted = isCompleted(productOrderVO);
+		if (isCompleted) {
+			log.debug("The received event is not in state completed.");
```

Comment:
Is the log wrong or am I misunderstanding the logic?

Reply:
The if check is wrong, thank you:)